src/openrct2/ride/Ride.cpp

```diff
@@ -816,18 +816,26 @@ bool track_block_get_previous(int32_t x, int32_t y, TileElement* tileElement, tr
  * bx result y
  * esi input / output map element
  */
-int32_t ride_find_track_gap(CoordsXYE* input, CoordsXYE* output)
+int32_t ride_find_track_gap(int32_t rideIndex, CoordsXYE* input, CoordsXYE* output)
 {
-    assert(input->element->GetType() == TILE_ELEMENT_TYPE_TRACK);
-    int32_t rideIndex = input->element->AsTrack()->GetRideIndex();
     Ride* ride = get_ride(rideIndex);
+    return ride_find_track_gap(ride, rideIndex, input, output);
+}
+
+int32_t ride_find_track_gap(Ride* ride, int32_t rideIndex, CoordsXYE* input, CoordsXYE* output)
```

Comment:
If you're already passing `rideIndex`, there is no need to pass in `Ride` as well. You can just do `get_ride` with that `rideIndex`.

Reply:
The point was to avoid using get_ride twice, most callers already have the Ride obtained, rideIndex is however still required by the function and there seems to be no way to get the ride index from the Ride struct.

Reply:
It seems unnecessary to pass both, though. `get_ride` doesn't take long to call (probably even gets inlined), or is this a step towards removing the index from calls, and having Rides store their own ID? If so, I'm not sure if that's really desirable, a value shouldn't care about its key.

Reply:
So what should be the final solution? Using get_ride over and over doesn't seem right either.

Reply:
Just do `Ride* ride = get_ride(ride_index)` at the start?

Reply:
Alright, that leaves janisozaur suggestion, apparently I can't have it both ways.